internal/db/collection_id.go

```diff
@@ -0,0 +1,656 @@
+						skippedSets = append(skippedSets, collectionSet)
+						continue setLoop
+					}
+				}
+			}
+		}
+
+		result = append(result, collectionSet)
+	}
+
+	if len(skippedSets) > 0 {
+		return sortCollectionSetsFrom(len(result), append(result, skippedSets...))
+	}
+
+	return result
+}
+
+// saveBlocks saves the collection set to the block and headstore.  It mutates the given collectionSet
+// setting the ids and migrations.
+func saveBlocks(
+	ctx context.Context,
+	collectionSet []*client.CollectionVersion,
+	migration immutable.Option[model.Lens],
+) error {
+	colIds := make([]cidlink.Link, 0, len(collectionSet))
+	hasSetUpdated := false
+
+	for _, collection := range collectionSet {
+		if collection.VersionID == "" && collection.CollectionID != "" {
+			// If the VersionID is empty, but the CollectionID is not, the user has patched one
+			// of these properties - continue, and let the validation code return an error.
+			continue
+		}
+
+		var oldCol client.CollectionVersion
+		if collection.VersionID != "" {
+			var err error
+			oldCol, err = description.GetCollectionByID(ctx, collection.VersionID)
+			if err != nil {
+				if errors.Is(err, corekv.ErrNotFound) {
+					// If the key does not exist, continue, and let the validation code handle it
+					// in a user friendly way.
+					continue
+				}
+				return err
+			}
+		}
+
+		var hasFieldsChanged bool
+		newFieldLevelCIDs := []coreblock.DAGLink{}
+		for i, newField := range collection.Fields {
+			fieldCRDT := crdt.NewFieldDefinition(collection.Name, newField.Name)
+			delta, hasFieldChanged, err := fieldCRDT.Delta(
+				newField,
+				// We cheat here for now, as users cannot yet mutate fields.  When they can,
+				// we will need to pass in the old version here.
+				client.CollectionFieldDescription{},
+			)
+			if err != nil {
+				return err
+			}
+
+			if !hasFieldChanged {
+				continue
+			}
+			hasFieldsChanged = true
+
+			cid, _, err := coreblock.AddDelta(ctx, fieldCRDT, delta)
+			if err != nil {
+				return err
+			}
+
+			collection.Fields[i].FieldID = cid.String()
+			newFieldLevelCIDs = append(newFieldLevelCIDs, coreblock.DAGLink{Link: cid})
+		}
+
+		colCRDT := crdt.NewCollectionDefinition(collection.Name)
+		delta, hasCollectionChanged := colCRDT.Delta(*collection, oldCol)
+
+		if !hasFieldsChanged && !hasCollectionChanged {
+			// If the global collection state has not changed, there is nothing to do here and we
+			// move on to the next collection.
+			continue
+		}
+		hasSetUpdated = true
+
+		cid, _, err := coreblock.AddDelta(ctx, colCRDT, delta, newFieldLevelCIDs...)
+		if err != nil {
+			return err
+		}
+
+		collection.VersionID = cid.String()
+		if collection.CollectionID == "" {
+			collection.CollectionID = collection.VersionID
+		}
+
+		colIds = append(colIds, cid)
+
+		newColSources := collection.CollectionSources()
+		oldColSources := oldCol.CollectionSources()
+		if oldCol.VersionID != "" && len(newColSources) == len(oldColSources) {
```

Comment:
question: I assume this implies that if this is not the root version and that this is not a view? Or maybe I'm getting this wrong. 

thought: It seems that for a normal collection, this can only ever be true on the first update. Subsequent updates would never have `len(newColSources) == len(oldColSources)` as true since the `len(newColSources)` will always be 0.

Reply:
question: Why is this done here? It seems irrelevant to the storage of the blocks.

Reply:
> question: Why is this done here? It seems irrelevant to the storage of the blocks.

It is where the collection ids are handled, it is the easiest place to do this as other locations would have to re-figure out what the old version was, duplicating a fair amount of CPU effort and code.

Reply:
> It will actually be the opposite - every update apart from the first, so long as the user has not manually patched the sources (or the system, which IIRC can happen for orphan collection versions, where a migration was added before the version was otherwise defined).

On the first update they would both be of length 0. I think what I'm missing is where the sources are set on the new collection for them not to be 0 on the next update.

Reply:
Fred and I spoke over Discord.  I believe he understands what and why the code is doing, but this block needs a fair bit of documentation.

- [ ] Document code block